teps/0122-reproducibility-of-complete-build-instructions.md

```diff
@@ -0,0 +1,298 @@
+
+### TaskRun Specification
+
+The table below provides the name and description of the API fields that the TaskRun requires. In addition, it also lists if that information is required, not required, or provided by the provenance. If it is required, it also shows where the information should be provided. Note that since task run provides information during runtime, most of the information is contained in the **invocation.Parameters** section of the provenance.
+
+
+| | **Metadata**      | | |
+| ------ | ------------- | ----- | ----- |
+| **Field Name**    | **Description** | **Not Required / Required / Provided by provenance** | **Insert in provenance** |
+| name| Name of the task run| Required (provide an audit trail)|  |
+| | **Spec** | | |
+| resources | resources used by task | Provided (can be extracted from steps/entrypoint) |  |
+| service account name | name of the service account | Required | buildConfig |
+| params | parameter values provided by taskrun | Provided |  |
+| workspaces | workspaces used by the task | Required | buildConfig |
+| pod template | | Required | buildConfig |
+| Timeout | time in which a task should complete | Not Required (can probably use default?) |  |
+| StepOverrides | Override Step configuration specified in a Task. Currently we can override compute resources. | Not Required |  |
+| SidecarOverrides | Override Sidecar configuration specified in a Task. Currently we can override compute resources. | Not Required |  |
+| ComputeResources | Configure compute resources required by the steps in the task. | Required (because this could also indicate a minimum amount of resources required for the task run.) | buildConfig |
+| [TaskSpec](#task-specification) | Specification of the resolved task | See [spec](#task-specification) | buildConfig |
+| [TaskRef](#taskref)  | Details of the referenced task | Not Required (Since we save the complete TaskSpec even that of a remote task) |  |
+| | **Status** | | |
+| Task Results | Results produced by the task run | Not Required | buildConfig |
+
+### Tekton System Versioning
+The **version (uri and commit-sha)** of **Tekton pipelines** (that reconciled the task run), **Tekton chains** (that produced the provenance from the task run) and **remote resolvers**. Capturing this information is vital for reproducibility of the task run and can be stored in **invocation.environment** section of the provenance.
+
+In order to provide this information, cluster admins can use a `configMap` in the **tekton-chains namespace** to provide versions of the installed Tekton systems. That way, chains can access the information easily for each task/pipeline run resolved using the installed system.
```

Comment:
Not sure if a configMap is the best place for this - there's no guarantee that this matches what was actually done, and you might be in situations where you're running mixed versions (e.g. during rollouts) where this may be ambiguous.

For Pipelines, we already include some information about the Run object itself about the reconciler that handled it, though we may want to improve this - https://github.com/tektoncd/pipeline/issues/5303

For Chains we can derive this in the chains reconciler itself.

For remote resolvers we should derive this from the ConfigSource.

Reply:
Good point!

I will capture the points regarding pipelines and chains. I'm wondering how we can derive the remote resolver version from the ConfigSource. Just like the chains reconciler, can the remote resolver reconciler attach its version to the resolution request CRD? That way, it can be forwarded to TaskRun's CRD and Chains can get it eventually?

Reply:
Does anyone know what ended up happening with [TEP-0041](https://github.com/tektoncd/community/blob/main/teps/0041-tekton-component-versioning.md)? B/c that seems to be treading very similar ground as this. ( @vinamra28 @vdemeester @nikhil-thomas @afrittoli ) It seems like our current approach is to add a label to the Deployment with the version (https://github.com/tektoncd/pipeline/blob/b3e4f35712fca0235e3ef80cc3f03b2600f40cb1/config/controller.yaml#L27) but I'm guessing Chains won't have permission to read these values

I suggest that we need to make sure this info is automatically provided as part of installing Tekton components or this won't work b/c it would be very error prone to expect admins to keep something like this up to date manually.

Reply:
Going through it, I was wondering if we can make system-versioning out-of-scope but include `configuration-feature-flags`?

Reply:
> Does anyone know what ended up happening with [TEP-0041](https://github.com/tektoncd/community/blob/main/teps/0041-tekton-component-versioning.md)? B/c that seems to be treading very similar ground as this. ( @vinamra28 @vdemeester @nikhil-thomas @afrittoli ) It seems like our current approach is to add a label to the Deployment with the version (https://github.com/tektoncd/pipeline/blob/b3e4f35712fca0235e3ef80cc3f03b2600f40cb1/config/controller.yaml#L27) but I'm guessing Chains won't have permission to read these values
> 
> I suggest that we need to make sure this info is automatically provided as part of installing Tekton components or this won't work b/c it would be very error prone to expect admins to keep something like this up to date manually.

@bobcatfish yeah it's implemented in most of the tekton components have this configmap along with appropriate RBACs. I believe chains will have to that ConfigMap and can easily fetch the version of the component that is installed. You can probably refer to the PR https://github.com/tektoncd/pipeline/pull/3971